### ISSUE TITLE
Addon occasionally stops working

### ISSUE THREAD

**Issue by kuchitsu** [birtles/rikaichamp#17]
Rikaichamp occasionally stops working for me on latest Firefox. I point at a Japanese text and nothing happens anymore. Re-enabling the addon doesn't fix the problem, but restarting the browser always does. At this point I haven't noticed anything in particular that results in it breaking, it seems to sort of just happen. Please tell me if there is some way for me to debug this and perhaps collect more information.

**Comment by birtles** [comment]
That's a real worry. I haven't encountered that yet. The most helpful thing would be to open the Browser Console (Ctrl+Shift+J or go to the Menu on the right -> Web Developer -> Browser Console) and see if you can see any errors from rikaichamp. If you filter on 'rikai' it should probably turn up most errors from rikaichamp (unless they came from options.js, data.js, or config.js -- so you might want to try filtering on those terms too).

Thanks for helping to debug this!

**Comment by kuchitsu** [comment]
The only seemingly related message I've seen in the console so far is this:

> Error: Could not establish connection. Receiving end does not exist.

It appears when I click on the Rikaichamp icon in order to disable it (no matter if it already stopped working or not).

Also I feel like it always takes at least a couple of hours before it stops working, but I'm not certain at this moment.

**Comment by birtles** [comment]
Hmm, that's weird. I see that message when I develop the extension but I think it happens then when the extension is re-loaded. If that message is related then I guess it means either the content part or chrome part has died. Perhaps the chrome part gets evicted because it consumes too much memory. But I'm just guessing. If that were the case then moving the database to IndexedDB would fix it.

Perhaps I've never encountered it because the machines I use have more memory, or I hoard less tabs, or because I only turn on Rikaichamp occasionally and then turn it off as soon as I've looked up the word I didn't know.

If you find any more things that hint at what triggers this that would be helpful (e.g. did you have a lot of tabs open at the time, was Firefox using a lot of memory etc.).

**Comment by kuchitsu** [comment]
I usually have very few tabs open, but this computer is rather old (only 2GB memory) so overload is possible I guess. I wish I could give more information but unfortunately I haven't really noticed any patterns or anything.

**Comment by birtles** [comment]
I'm really sorry you're still experiencing this. I have started working on the alternative approach that should require less memory but it will be quite some time before it is ready, I'm afraid.

Until then, if you notice any other patterns that lead to this, please let me know!

**Comment by Esoppant** [comment]
I've been having the exact same problem, and yet my PC has 8 GBs of RAM. So it might not be a RAM problem.

It tends to happen after I leave Firefox unclosed for, approximately more than 12-18 hours, and get fixed with a restart of Firefox. Other than that, I haven't noticed any patterns too, sadly.

On the console, I can see these messages (although since they are not timestamped, I don't know when exactly they are from):

> Loading failed for the <script> with source “blob:https://outgoing.prod.mozaws.net/8c3a5bcc-d3dc-4658-b1a3-682961045490”.  rikaichamp:1
> 
> Content Security Policy: The page’s settings blocked the loading of a resource at self (“script-src 'none'”). Source: 
>       
> 
>       setTimeout(function() {
>  ....  rikaichamp:42
> 
> Loading failed for the <script> with source “blob:https://github.com/fac2180c-9a29-465c-8f4e-94862722fbb6”.  rikaichamp:1

I use NoScript and I think the first and last messages have to do with NoScript blocking some scripts while getting to this github page. Not sure what the message in the middle is.

**Comment by birtles** [comment]
This seems to be happening more frequently now. I'd like to debug it but I'm not quite sure where to start.

If anyone encounters it would, it would help if you could provide the following information:

* Version of Firefox you were using
* How long you had it open
* Was this the first time you enabled rikaichamp in this session?
* What page were you viewing?
* Was the page still loading when you enabled rikaichamp?

**Comment by kikaxa** [comment]
Have encountered something similar once, usually I enable the addon occasionally on use, but that time i had it enabled for ~1h browsing session on accident(maybe since ff restart?), and the next time it did nothing. No messages in console, though. ff 59, first time.

Had multiple strange cases when enabling addon during loading\back-navigation, and it worked/not worked simultaneously in other tabs/history pages. Have not checked logs, though.

**Comment by kikaxa** [comment]
Had more occurences of this today on youtube, have tested it a litte bit.
It definitely looks like some kind of race condition or incorrect interruption:

I have mouse with additional back/forward buttons that i use extensively.
My semi-reproducible steps for this bug:
(It might have helped that i had a large torrent/download in background at that time. maybe network limit in dev tools will work for this.)
- Open  youtube video. Navigate to the next one. Restart rikaichan for a clean state.
- Click back. Scroll a bit till comments start loading.
While all this stuff is loading:
- Move mouse rapidly over japanese text(i have japanese locale so there are plenty), click forward mid-movement.

While still loading new page: 
- Scroll a bit till comments start loading. At this time move mouse over japanese description.
- Now scroll 1-2screens down, try using rikaichan there on the comments.

At this point i sometimes(30-50%) get it: 
- Rikai finds text only from the description
- Rikai works only when scrolled to top
- Stops/starts working after a while/some scrolling and mouse movement
- Popup freezes somewhere

**Comment by lae** [comment]
@birtles regarding the switching over to IndexedDB comment - the implementation on Firefox (and maybe Chromium, have not checked) has a wildly greedy expectation with regards to disk space usage, where it will delete a database on startup if it finds that there isn't 10x as much space available as is required, and is the primary reason why I was forced to stop using Yomichan and use rikaichamp. See https://github.com/FooSoft/yomichan/pull/77#issuecomment-329037070

(this issue also affects me)

**Comment by birtles** [comment]
@kikaxa Thank you for the detailed steps! I will give it a try next week and see what I can figure out.

@lae Thank you for the very helpful info! It looks like this is [documented on MDN](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Browser_storage_limits_and_eviction_criteria#Storage_limits) (where 20% of 50% matches the 10% you reported). I suppose the current approach avoids that so maybe IndexedDB is not so great.

**Comment by birtles** [comment]
I tried some of these steps and I can reproduce some of the symptoms. As best I can tell, however, it just seems to be a performance issue (as opposed to the add-on hitting some error state).

The fact that some people find disabling and re-enabling the add-on fixes it might suggest that the background script is flooded with lookup requests.

I've also noticed the add-on performs poorly on facebook after scrolling for some time.

At this stage, the best mitigation strategies I can think of are:

* Profiling the content script and trying to optimize the selection lookup (trying to linearize the cost when working on longer pages -- especially infinite/progressive scrollers like facebook / YouTube comments).
* Doing some debouncing / aborting of lookups.
* Switching to IndexedDB for lookups (might relieve memory pressure but @lae points out this might introduce issues on machines with limited disk space).
* Moving the lookup to a worker?

**Comment by birtles** [comment]
Trying a bit more I managed to reproduce the issue with the stuck popup. It happened for me when browsing Youtube with the Browser Console open. Youtube spews a bunch of console messages (on Firefox 60 at least) which, as far as I can tell, ties down the parent process (or at least the content script  ↔ background script communication).

I believe this makes it easier to get to a situation where there are overlapping dictionary lookup requests shortly after page load. I suspect that can subsequently cause us to get us into a situation where we have overlapping calls to `showPopup` which blocks on the stylesheet load, hence we end up creating multiple elements with ID `rikaichamp-window` (since the ID is assigned after the stylesheet load has completed). Basically, the first concrete step to fix this is just to make the content script better handle overlapping dictionary lookups.

**Comment by Lebon14** [comment]
Yeah, it also happens to me too. I usually notice it when using Twitter. A page refresh usually fixes it.

**Comment by birtles** [comment]
I wonder if this has improved at all with version 0.0.14 (released ~10 days ago). It should have fixed one potential source of this bug.

**Comment by Esoppant** [comment]
It is likely that the bug is less frequent now. I don't recall having this problem within the last week.

**Comment by Lebon14** [comment]
I still get the issue. At least I remember it happening.

**Comment by maranyc** [comment]
Me too, quite annoying....happens on the site where I study Kanji - wanikani.....haven't looked at that many other Japanese sites lately...

**Comment by maranyc** [comment]
seems like it's okay today....yay!  Don't know why but I'm happy. :)

**Comment by maranyc** [comment]
Me again....just checked the Firefox add-ons page and saw it was updated yesterday....so maybe that's how come it's working today.  Thank you very much!

**Comment by birtles** [comment]
Ok, I might have finally worked this out. Here's my theory anyway. We have the following code:

In `shouldThrottlePopup` we have the following code:

```js
  const now = performance.now();
  let averageSpeed = 0;

  if (this._previousMousePosition && this._previousMouseMoveTime) {
    const distance = Math.sqrt(
      Math.pow(ev.pageX - this._previousMousePosition.x, 2) +
        Math.pow(ev.pageY - this._previousMousePosition.y, 2)
    );
    const speed = distance / (now - this._previousMouseMoveTime);

    this._mouseSpeeds.push(speed);
    this._mouseSpeedRollingSum += speed;

    if (this._mouseSpeeds.length > RikaiContent.MOUSE_SPEED_SAMPLES) {
      this._mouseSpeedRollingSum -= this._mouseSpeeds.shift()!;
    }

    averageSpeed = this._mouseSpeedRollingSum / this._mouseSpeeds.length;
  }

  this._previousMousePosition = { x: ev.pageX, y: ev.pageY };
  this._previousMouseMoveTime = now;

  return averageSpeed >= RikaiContent.MOUSE_SPEED_THRESHOLD;
```

If this returns false we'll hide the popup.

Now, suppose that the browser is under load and our calls to this method from `onMouseMove` get backed up such that they happen at the same time as reported by `performance.now()`―something that became more likely since Firefox started reducing the resolution of `performance.now()` earlier this year.

That could produce a situation where `now` is equal to `this._previousMouseMoveTime()`. As a result we get:

```js
const speed = distance / (now - this._previousMouseMoveTime);
```

which gives `speed = distance / 0 = Infinity`.

Then we add that to the rolling sum:

```
this._mouseSpeeds.push(speed);
this._mouseSpeedRollingSum += speed;
```

Such that `this._mouseSpeedRollingSum` becomes `Infinity`.

Now, the problem with that is `Infinity - x === Infinity` for most `x` so that even when we try to drop this sample later, the rolling sum will remain `Infinity`. Effectively `this._mouseSpeedRollingSum` will be `Infinity` forever.

Then we have:

```js
averageSpeed = this._mouseSpeedRollingSum / this._mouseSpeeds.length;
```

And since `Infinity / x === Infinity` for most `x`, `averageSpeed` will also be `Infinity`.

Finally, when we go to evaluate:

```
return averageSpeed >= RikaiContent.MOUSE_SPEED_THRESHOLD;
```

It will always return true and so we will always hide the popup.

I think that explains the bug and also explains why this became more frequent earlier this year. Thanks to @maranyc for giving me the hint I needed to find this!

I will try to fix this by:

* Switching to event times
* Checking for when the event times are exactly equal and probably just dropping the new sample and perhaps updating the old sample in that case

**Comment by birtles** [comment]
Oh, I don't know how this got through, but apparently I really messed up when I fixed the timer precision stuff last time. It looks like I made it do the exact opposite of what it was supposed to do. I don't know why or how it worked for anyone.

**Comment by maranyc** [comment]
Well, I haven't the foggiest idea what you are talking about, Mr. birtles....but if I somehow helped you I am pleased and it is still working for me!

**Comment by mjkarlin** [comment]
I'm not a programmer, but I have used rikaichamp for quite a while.  As of yesterday (11/26/2020), it stopped working in Firefox.  Like other users, I have disabled and re-enabled it and I have tried restarting Firefox and even rebooting.  I have tried it with various Japanese language websites as well as Wanikani.  No luck.  Nothing came up for rikai on Ctrl-Shift-J.  I know it's Thanksgiving weekend, so no rush, but any help would be appreciated.
